keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,152 @@
+
+### Goals
+
+* Provide a common way to request common services (e.g., MySQL)
+* Provide a common means to obtain credentials to use the service
+* Provide a common method to detect which services are available
+* Provide a system that can be implemented for the major public clouds, on-premise clusters, and local cluster (e.g., Docker for Mac)
+
+### Non-Goals
+
+* Provide an out of the box solution for every bespoke service provided by everyone
+* Replace Kubernetes Service Catalog
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user of Kubernetes, I can query the services I can declaratively request using a Kubernetes native API. For example, using the command `kubectl get crds` where I can see services alongside other resources that can be created.
+
+#### Story 2
+
+As a user of Kubernetes, I can declaratively request an instance of a service using a custom resource. When the service is provided the means to use that service (e.g., credentials in a secret) are provided in a common and consistent manner. The same resource and secret can be used in clusters running in different locations and the way the service is provided may be different.
+
+#### Story 3
+
+As a cluster operator or application operator, I can discover controllers implementing the CRDs and secrets to support the application portability in my cluster.
+
+#### Story 4
+
+As a cluster operator or application operator, I can set default values and provider custom settings for a service. 
+
+### Implementation Details/Notes/Constraints
+
+To solve the two user stories there are two types of Kubernetes resources that can be leveraged.
+
+1. Custom resource definitions (CRDs) can be used to describe a service. The CRDs can be implemented by controllers for different environments and the list of installed CRDs can be queried to see what is supported in a cluster
+2. Secrets with a specific type and schema can be used to handle credentials and other relevant information for services that have them (e.g., a database). Not all services will require a secret (e.g., DNS)
+
+This subproject will list and document the resources and how controllers can implement them. This provides for interoperability including that for controllers and other tools, like validators, and a canonical listing held by a vendor neutral party.
+
+In addition to the resources, this subproject will also provide a controller implementing the defined services to support testing, providing an example implementation, and to support other Kubernetes subprojects (e.g., Minikube). Controllers produced by this project are _not_ meant to be used in production.
+
+3rd party controllers implementing the CRDs and secrets can use a variety of methods to implement the service handling. This is where the Kubernetes Service Catalog can be an option. This subproject will not host or support 3rd party controllers but will list them to aide in users discovering them. This is in support of the 3rd user story.
```

Comment:
You think it would be a good idea to co-location implementations for AWS, Azure, Alibaba, Google Cloud, Oracle Cloud, IBM Cloud, and a dozen others in one place? Oh, and what if two groups come up with different implementation for a single cloud provider? Then there are other situations such as DNS via cloudflare and all the other service provider. What about them? By putting hosting out of scope it's a problem we don't have to solve.